### PR TITLE
Improving DataModelOracle: inspecting generic types on class fields.

### PR DESCRIPTION
**Pull request by pefernan**

### CODE REVIEW

File: kie-wb-common-services/kie-wb-common-datamodel/kie-wb-common-datamodel-backend/src/test/java/org/kie/workbench/common/services/datamodel/backend/server/ProjectDataModelFactFieldsGenericsTest.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package org.kie.workbench.common.services.datamodel.backend.server;
```
Comment (manstis) [inline kie-wb-common-services/kie-wb-common-datamodel/kie-wb-common-datamodel-backend/src/test/java/org/kie/workbench/common/services/datamodel/backend/server/ProjectDataModelFactFieldsGenericsTest.java]: License is missing.

File: kie-wb-common-services/kie-wb-common-datamodel/kie-wb-common-datamodel-backend/src/test/java/org/kie/workbench/common/services/datamodel/backend/server/testclasses/ProductOrder.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package org.kie.workbench.common.services.datamodel.backend.server.testclasses;
```
Comment (manstis) [inline kie-wb-common-services/kie-wb-common-datamodel/kie-wb-common-datamodel-backend/src/test/java/org/kie/workbench/common/services/datamodel/backend/server/testclasses/ProductOrder.java]: License is missing.

**Comment by manstis** [comment]
@pefernan Just a couple of missing licenses; and code formatting looks wrong.. but changes themselves looks fine. Could you please add the licenses and check your code formatter (the code in this PR is formatted differently to, for example, your commits on LiveSpark).

**Comment by pefernan** [comment]
@manstis curious, I'm using the same profile it should take the same formatting template.

**Comment by manstis** [comment]
+1 to merge. Thanks @pefernan

File: kie-wb-common-services/kie-wb-common-datamodel/kie-wb-common-datamodel-backend/src/main/java/org/kie/workbench/common/services/datamodel/backend/server/builder/projects/ClassFactBuilder.java [hunk 3]
```diff
@@ -173,6 +180,27 @@ private void loadClassFields( final Class<?> clazz,
...
     }
 
+
+    protected void discoverFieldFactBuilder( final String genericTypeName,
+                                             final Class<?> genericType,
+                                             final Map<String, FactBuilder> discoveredFieldFactBuilders ) throws IOException {
+
+        if ( !discoveredFieldFactBuilders.containsKey( genericTypeName ) ) {
+            discoveredFieldFactBuilders.put( genericTypeName, null );
+            discoveredFieldFactBuilders.put( genericTypeName,
```
Comment (yurloc) [inline kie-wb-common-services/kie-wb-common-datamodel/kie-wb-common-datamodel-backend/src/main/java/org/kie/workbench/common/services/datamodel/backend/server/builder/projects/ClassFactBuilder.java]: Seems like line 189 is redundant.

File: kie-wb-common-services/kie-wb-common-datamodel/kie-wb-common-datamodel-backend/src/main/java/org/kie/workbench/common/services/datamodel/backend/server/builder/projects/ClassFactBuilder.java [hunk 4]
```diff
@@ -173,6 +180,27 @@ private void loadClassFields( final Class<?> clazz,
...
+                                             new ClassFactBuilder( builder,
+                                                                   discoveredFieldFactBuilders,
+                                                                   genericType,
+                                                                   false,
+                                                                   typeSource ) );
+
+        }
+        if ( discoveredFieldFactBuilders.get( genericTypeName ) != null ) {
```
Comment (yurloc) [inline kie-wb-common-services/kie-wb-common-datamodel/kie-wb-common-datamodel-backend/src/main/java/org/kie/workbench/common/services/datamodel/backend/server/builder/projects/ClassFactBuilder.java]: This null check is redundant, isn't it?

**Comment by yurloc** [comment]
So I tried to remove the seemingly redundant calls and learned the hard way that they are not redundant at all. Thanks for very good code coverage! The piece of code I was looking at is just a little bit cryptic.

**Pull request merged**